kura/org.eclipse.kura.container.orchestration.provider/src/main/java/org/eclipse/kura/container/orchestration/provider/impl/DockerServiceImpl.java

```diff
@@ -271,97 +245,141 @@ public String getContainerIDbyName(String name) {
             String[] containerNames = cont.getNames();
             for (String containerName : containerNames) {
                 if (containerName.equals("/" + name)) { // docker API seems to put a '/' in front of the names
-                    return cont.getId();
+                    return Optional.of(cont.getId());
                 }
             }
         }
 
-        return "";
+        return Optional.empty();
     }
 
     @Override
-    public void stopContainer(String id) {
+    public void startContainer(String id) throws KuraException {
         if (!testConnection()) {
             throw new IllegalStateException(UNABLE_TO_CONNECT_TO_DOCKER_CLI);
         }
-        this.dockerClient.stopContainerCmd(id).exec();
+        try {
+            this.dockerClient.startContainerCmd(id).exec();
+        } catch (Exception e) {
+            logger.error("Could not start container {}. It could be already running or not exist at all", id);
+            throw new KuraException(KuraErrorCode.OS_COMMAND_ERROR);
+        }
     }
 
     @Override
-    public void startContainer(String id) {
+    public void startContainer(ContainerDescriptor container) throws KuraException {
+        if (isNull(container)) {
+            throw new IllegalArgumentException(CONTAINER_DESCRIPTOR_CANNOT_BE_NULL);
+        }
         if (!testConnection()) {
-            throw new IllegalStateException(UNABLE_TO_CONNECT_TO_DOCKER_CLI);
+            throw new KuraException(KuraErrorCode.CONNECTION_FAILED, "Unable to connect to the container service");
         }
-        try {
-            this.dockerClient.startContainerCmd(id).exec();
-        } catch (Exception e) {
-            logger.error("Could not start container {}. It could be already running or not exist at all", id);
+
+        logger.info("Starting {} Microservice", container.getContainerName());
+
+        Optional<String> containerId = getContainerIdByName(container.getContainerName());
+
+        if (!containerId.isPresent()) {
+            pullImage(container.getContainerImage(), container.getContainerImageTag(), 120);
+            containerId = createContainer(container);
         }
+
+        startContainer(containerId.orElse(""));
+
+        logger.info("Container Started Successfully");
+
     }
 
     @Override
-    public void deleteContainer(String id) {
+    public void stopContainer(String id) {
         if (!testConnection()) {
             throw new IllegalStateException(UNABLE_TO_CONNECT_TO_DOCKER_CLI);
         }
-        this.dockerClient.removeContainerCmd(id).exec();
+        this.dockerClient.stopContainerCmd(id).exec();
     }
 
     @Override
-    public void pullImage(String imageName) {
-        pullImage(imageName, 120);
+    public void stopContainer(ContainerDescriptor container) {
```

Comment:
What about just removing this method? The only field that it consider in the `ContainerDescriptor` is the container name (previously the container id). 
Probably the `stopContainer(String id)` is the minimum required functionality here, the clients can figure out the id given a name/container descriptor using the other methods (this is what the current implementation does). The same applies to `deleteContainer(ContainerDescriptor containerDescriptor)`.